### PR TITLE
[12.0] base: Fix search counts

### PR DESCRIPTION
**Pull request by ddejong-therp**
**Description of the issue/feature this PR addresses:**
When a search call is made, sometimes the number of records that are being returned are less than the limit being set.

**Current behavior before PR:**
Currently, when a sql search query is constructed, it 'uniqifies' the result set in python, but the limit is set within sql.
So basically the limit is being implemented in a record set which includes duplicates.
So: unique-record-count <= limit.
Filters sometimes make use of One2many fields, and those search calls get turned into sql queries which can return duplicate ID's of the table/model you are trying to filter.

However, when the actual records are not needed and only the count of the records are, the records are counted in sql, and not python. So this returns a different count than records that would be returned, and this causes issues further down the line.

For example, when a filter is applied that returns a count of 89, but gives only 66 actual records, 23 records are not shown in the (tree) view with a limit of 80.

**Desired behavior after PR is merged:**
The search ORM calls return the correct number of records, and the correct record counts.

This is done by making use of SQL's DISTINCT function, so that the ID list is uniqified before the limit is applied.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ddejong-therp** [comment]
This is basically the same thing as this PR done here: https://github.com/OCA/OCB/pull/841
But that one is for odoo 10, this is for 12.

We both take slightly different approach. However, I think that moving the burden of uniquifying the resultset from python to sql, probably causes less overhead (if any) than wrapping your query with another query, as is done with that other PR.

File: odoo/models.py [hunk 1]
```diff
@@ -4093,23 +4094,27 @@ def _generate_order_by_inner(self, alias, order_spec, query, reverse_direction=F
...
+                order_by_fields.append('"%s"."%s"' % (alias, order_field))
             else:
                 if field.inherited:
                     field = field.base_field
                 if field.store and field.type == 'many2one':
                     key = (field.model_name, field.comodel_name, order_field)
                     if key not in seen:
                         seen.add(key)
-                        order_by_elements += self._generate_m2o_order_by(alias, order_field, query, do_reverse, seen)
+                        els, fields = self._generate_m2o_order_by(alias, order_field, query, do_reverse, seen)
```
Comment (NL66278) [inline odoo/models.py:4105]: I suppose `els` stands for `elements`? Avoid random abbreviations in variable names (or any other names).

**Comment by NL66278** [comment]
Did you submit this to the main Odoo branch as well? Would be nice if Odoo SA included your fix.

**Comment by ddejong-therp** [comment]
No I did not submit it to the main Odoo branch yet, but I will do that now.

File: odoo/models.py (hunk 1 above)
Comment (ddejong-therp) [reply odoo/models.py:4105]: Alright, ill take that advice and change it. :)

**Pull request closed without merging**